File: LeetCode/Letter_Combinations_of_a_Phone_Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        dic = {'2': ['a', 'b', 'c'],
               '3': ['d', 'e', 'f'],
               '4': ['g', 'h', 'i'],
               '5': ['j', 'k', 'l'],
               '6': ['m', 'n', 'o'],
               '7': ['p', 'q', 'r', 's'],
               '8': ['t', 'u', 'v'],
               '9': ['w', 'x', 'y', 'z']}
        
        result = []
        def helperLetterComb(comb, nextDigits):
            if len(nextDigits) == 0:
                result.append(comb)
            else:
                for letter in dic[nextDigits[0]]:
                    helperLetterComb(comb + letter, nextDigits[1:])
                    
        if digits:
            helperLetterComb("", digits)
        return result

File: LeetCode/Letter_Combinations_of_a_Phone_Number.cpp

class Solution {
public:
    std::vector<string> letterCombinations(string digits) {
        std::vector<std::string> combs = {};
        std::vector<std::vector<char>> phoneTable = {
            {' ', ' ', ' '},
            {' ', ' ', ' '},
            {'a', 'b', 'c'},
            {'d', 'e', 'f'},
            {'g', 'h', 'i'},
            {'j', 'k', 'l'},
            {'m', 'n', 'o'},
            {'p', 'q', 'r', 's'},
            {'t', 'u', 'v'},
            {'w', 'x', 'y', 'z'},
        };
        int index = 0;
        int c = 0;
        while(index != digits.size()) {
            if(c == 100)
                break;
            std::vector<char> currentDigitEntry = phoneTable[digits[index] - 48];
            if(combs.size() == 0) {
                for(auto k : currentDigitEntry) {
                    std::string s(1, k);
                    combs.push_back(s);
                }
                index++;
                continue;
            }
            for(int i = 0 ; i < currentDigitEntry.size() ; i++) {
                int size = combs.size();
                for(int j = 0 ; j < size ; j++) {
                    
                    combs.push_back(combs[j] + currentDigitEntry[i]);
                }
            }
            index++;
        }
        for(int i = combs.size() - 1; i >= 0 ; i--) {
            if(combs[i].size() != digits.size()) {
                combs.erase(combs.begin() + i);
            }
        }
        return combs;
    }
};
